cpp/src/bitmask/bitmask_util.cuh

```diff
@@ -0,0 +1,324 @@
+//			Device Only Utility Functions
+//----------------------------------------------------------------------------------------------------------
+namespace device {
+
+	/**
+	 * Check to see if a record is Valid (aka not null)
+	 *
+	 * @param[in] valid        the device memory containing the valid bitmaps
+	 * @param[in] record_idx   the record index to check
+	 *
+	 * @return  true if record is valid, false if record is null
+	 */
+	__device__ bool is_valid(gdf_valid_type * valid, int record_idx) {
+		int rec = bitmask::util::which_bitmap_record(record_idx);
+		int bit = bitmask::util::which_bit(record_idx);
+
+		int status = atomicAnd(&valid[rec], (1U << bit));
+
+		return ( status == 0) ? false : true;
+
+	}
+
+
+	/**
+	 * Set a bit
+	 *
+	 * @param[in] valid         the valid memory array
+	 * @param[in] record_idx    the record index
+	 *
+	 * @return gdf_error - did it work or not
+	 */
+	__device__ gdf_error set_bit(gdf_valid_type * valid, int record_idx) {
+		int rec = bitmask::util::which_bitmap_record(record_idx);
+		int bit = bitmask::util::which_bit(record_idx);
+
+		atomicOr( &valid[rec],   (1U << bit));		// set the bit
+
+		return ( cudaPeekAtLastError() == cudaSuccess) ? GDF_SUCCESS : GDF_CUDA_ERROR;
+	}
+
+
+	/**
+	 * Clear a bit
+	 *
+	 * @param[in] valid         the valid memory array
+	 * @param[in] record_idx    the record index
+	 *
+	 * @return gdf_error - did it work or not
+	 */
+	__device__ gdf_error clear_bit(gdf_valid_type * valid, int record_idx) {
+		int rec = bitmask::util::which_bitmap_record(record_idx);
+		int bit = bitmask::util::which_bit(record_idx);
+
+		atomicAnd( &valid[rec], ~(1U << bit));		// clear the bit
+
+		return ( cudaPeekAtLastError() == cudaSuccess) ? GDF_SUCCESS : GDF_CUDA_ERROR;
+	}
+
+} // end of device namespace
+
+
+//----------------------------------------------------------------------------------------------------------
+//		Utility Functions that launch kernels
+//----------------------------------------------------------------------------------------------------------
+namespace kernel {
+
+
+
+
+
+}
+
+
+
+
+//----------------------------------------------------------------------------------------------------------
+//			Host Utility Functions
+//----------------------------------------------------------------------------------------------------------
+namespace host {
+	/**
+	 * Allocate device space for the valid bitmap.
+	 *
+	 * @param[out] gdf_valid_type *      pointer to where device memory will be allocated and returned
+	 * @param[in]  number_of_records     number of records
+	 * @param[in]  fill_value            optional, should the memory be initialized to all 0 or 1s. All other values indicate un-initialized
+	 * @return error status
+	 */
+	gdf_valid_type * create_bitmap(int number_of_records, int fill_value = -1) {
+
+		gdf_valid_type *valid_d;
+
+		int num_bitmasks = (number_of_records + (GDF_VALID_BITSIZE - 1)) / GDF_VALID_BITSIZE;
```

Comment:
Note that this can lead to overflow if `number_of_records > (std::numeric_limits<int>::max() - GDF_VALID_BITSIZE)`

Reply:
Furthermore, this division with rounding up operation is performed in several places. It should be put into it's own function that is reused instead of redefining it many times.